src/Modules/TasksModule/components/TaskData/TaskData.tsx

```diff
@@ -1,5 +1,301 @@
-import React from "react";
-
+import axios from "axios";
+import { useEffect, useState } from "react";
+import { SubmitHandler, useForm } from "react-hook-form";
+import { useLocation, useNavigate } from "react-router-dom";
+import { toast } from "react-toastify";
+import {
+  ProjectInterface,
+  TaskSubmitInterface,
+  TaskUpdateSubmitInterface,
+  UsersInterface,
+} from "../../../../interfaces/Auth";
+import {
+  baseUrl,
+  handleApiError,
+  getRequestHeaders,
+} from "../../../../utils/Utils";
 export default function TaskData() {
-  return <>TaskData</>;
+  const [isLoading, setIsLoading] = useState(false);
+  const [pageNumber] = useState(1);
+  const [projects, setprojects] = useState([]);
+  const [users, setUsers] = useState([]);
+  const [isUpdate, setIsUpdate] = useState(false);
+  const [taskId, setTaskId] = useState(0);
+  const navigate = useNavigate();
+  const location = useLocation();
+  const task = location.state;
+  // console.log(location);
+  const {
+    register,
+    handleSubmit,
+    formState: { errors },
+    reset,
+  } = useForm<TaskSubmitInterface>({
+    criteriaMode: "all",
+  });
+
+  const onSubmit: SubmitHandler<TaskSubmitInterface> = async (task) => {
+    try {
+      await axios.post(`${baseUrl}/Task`, task, {
+        headers: getRequestHeaders(),
+      });
+      toast.success("New Task Has Been Added Successfully");
+      navigate("/dashboard/tasks");
+    } catch (error) {
+      handleApiError(error);
+    }
+  };
+
+  const getProject = async (pageSize: number) => {
+    setIsLoading(true);
+    try {
+      const { data } = await axios.get(
+        `${baseUrl}/Project?pageSize=${pageSize}&pageNumbe=${pageNumber}`,
+        {
+          headers: getRequestHeaders(),
+          // params: {
+          //   name: name,
+          // },
+        }
+      );
+      setprojects(data.data);
+    } catch (err) {
+      handleApiError(err);
+    }
+    setIsLoading(false);
+  };
+  const onUpdateSubmit: SubmitHandler<TaskUpdateSubmitInterface> = async (
+    data
+  ) => {
+    try {
+      await axios.put(`${baseUrl}/Task/${taskId}`, data, {
+        headers: getRequestHeaders(),
+      });
+      // handleCloseUpdate();
+      // getTask(taskName, 10, 1);
+      toast.success(`Updated Task Successfully`);
+      navigate("/dashboard/tasks");
+    } catch (error) {
+      handleApiError(error);
+    }
+    // console.log(data);
+  };
+  const getUsers = async (pageSize: number) => {
```

Comment:
The current implementation is okay, but you'll notice that we'll have duplicate codes in multiple places. We will need the same function in the users Pages. The same goes with getProjects. We can make a custom hook that will contain all the fetching logic (data,isLoading,isError ... etc) and use it when need for example: useUsers, useProjects. This will make our code more cleaner.

Reply:
You can watch this talk about custom hooks. I remember when I first watched it, I started adopting the concept of custom hooks and I really liked it:
https://youtu.be/J-g9ZJha8FE?si=UOl33XO8Bg2qe88j

Reply:
You might consider this as a plus to implement in this project and start implementing it in the next one.